File: app/src/main/java/com/iyaselerehoboth/otrek/WelcomeScreenActivity.java
package com.iyaselerehoboth.otrek;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.WindowManager;
import android.widget.Toast;

import com.facebook.AccessToken;
import com.facebook.CallbackManager;
import com.facebook.FacebookCallback;
import com.facebook.FacebookException;
import com.facebook.login.LoginManager;
import com.facebook.login.LoginResult;
import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.common.api.ApiException;
import com.google.android.gms.tasks.Task;
import com.google.android.material.button.MaterialButton;
import com.google.firebase.auth.AuthCredential;
import com.google.firebase.auth.FacebookAuthProvider;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.GoogleAuthProvider;
import com.iyaselerehoboth.otrek.Database.SessionManager;

import java.util.Arrays;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class WelcomeScreenActivity extends AppCompatActivity {
    @BindView(R.id.btn_facebook_login)
    MaterialButton btn_facebook_login;

    @BindView(R.id.btn_email_login)
    MaterialButton btn_email_login;

    @BindView(R.id.btn_google_login)
    MaterialButton btn_google_login;

    private CallbackManager mCallbackManager;
    private GoogleSignInClient mGoogleSignInClient;
    SessionManager session;
    private FirebaseAuth mAuth;
    private static String TAG = "OTrek CHECK";
    private static final int RC_SIGN_IN = 9001;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_welcome_screen);
        ButterKnife.bind(this);
        session = new SessionManager(WelcomeScreenActivity.this);
        session.clearUserDetails();

        // In Activity's onCreate() for instance
        // To make the notification bar transparent.
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS, WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS);

        //Configure Google's sign in
        GoogleSignInOptions gso = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(getString(R.string.oauth2_web_client_id))
                .requestEmail()
                .build();
        mGoogleSignInClient = GoogleSignIn.getClient(this, gso);

        mAuth = FirebaseAuth.getInstance();

    }

    @Override
    protected void onStart() {
        super.onStart();
        // Check if user is signed in (non-null) and update UI accordingly.
        FirebaseUser currentUser = mAuth.getCurrentUser();
        if(currentUser != null){
            saveUserAndRedirect(currentUser);
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        // Check if user is signed in (non-null) and update UI accordingly.
        FirebaseUser currentUser = mAuth.getCurrentUser();
        if(currentUser != null){
            saveUserAndRedirect(currentUser);
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        FirebaseAuth.getInstance().signOut();
        session.clearUserDetails();
    }

    @OnClick(R.id.btn_email_login)
    public void emailLogin(){
        startActivity(new Intent(WelcomeScreenActivity.this, ProfilePageActivity.class));
    }

    @OnClick(R.id.btn_facebook_login)
    public void facebookLogin(){
        mCallbackManager = CallbackManager.Factory.create();
        LoginManager.getInstance().logInWithReadPermissions(WelcomeScreenActivity.this, Arrays.asList("email", "public_profile"));
        LoginManager.getInstance().registerCallback(mCallbackManager, new FacebookCallback<LoginResult>() {
            @Override
            public void onSuccess(LoginResult loginResult) {
                Log.d(TAG, "facebook:onSuccess" + loginResult);
                firebaseAuthWithFacebook(loginResult.getAccessToken());
            }

            @Override
            public void onCancel() {

            }

            @Override
            public void onError(FacebookException error) {
                Toast.makeText(WelcomeScreenActivity.this, "Error: " + error.getLocalizedMessage(), Toast.LENGTH_LONG).show();
            }
        });
    }

    @OnClick(R.id.btn_google_login)
    public void googleLogin(){
        Intent googleSignInIntent = mGoogleSignInClient.getSignInIntent();
        startActivityForResult(googleSignInIntent, RC_SIGN_IN);
    }

    private void firebaseAuthWithFacebook(AccessToken token) {
        Log.d(TAG, "firebaseAuthWithFacebook:" + token);

       AuthCredential credential = FacebookAuthProvider.getCredential(token.getToken());
        mAuth.signInWithCredential(credential)
                .addOnCompleteListener(this, task -> {
                    if (task.isSuccessful()) {
                        Log.d(TAG, "signInWithCredential:success");
                        FirebaseUser user = mAuth.getCurrentUser();
                        saveUserAndRedirect(user);
                    } else {
                        // If sign in fails, display a message to the user.
                        Log.w(TAG, "signInWithCredential:failure", task.getException());
                        Toast.makeText(WelcomeScreenActivity.this, "Authentication failed.",
                                Toast.LENGTH_SHORT).show();
                    }
                });
    }

    private void firebaseAuthWithGoogle(GoogleSignInAccount acct){
        Log.d(TAG, "firebaseAuthWithGoogle: " + acct.getId());

       AuthCredential credential = GoogleAuthProvider.getCredential(acct.getIdToken(), null);
        mAuth.signInWithCredential(credential)
                .addOnCompleteListener(this, task -> {
                    if(task.isSuccessful()){
                        Log.d(TAG, "signInWithCredential:success");
                        FirebaseUser user = mAuth.getCurrentUser();
                        saveUserAndRedirect(user);
                    }else{
                        Log.d(TAG, "signInWithCredential:failure", task.getException());
                        Toast.makeText(WelcomeScreenActivity.this, "Authentication failed.",
                                Toast.LENGTH_SHORT).show();
                    }
                });
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if(requestCode == RC_SIGN_IN && resultCode == RESULT_OK){
            //Response meant for Google.
            Task<GoogleSignInAccount> task = GoogleSignIn.getSignedInAccountFromIntent(data);
            try{
                //Google sign in was successful, authenticate with Firebase.
                GoogleSignInAccount account = task.getResult(ApiException.class);
                firebaseAuthWithGoogle(account);
            }catch (ApiException e){
                if(e.getStatusCode() == 7){
                    //Network Error. Retry with good network
                    Toast.makeText(this, "NETWORK ERROR, Kindly check your network and try again", Toast.LENGTH_LONG).show();
                }
                Log.d(TAG, "Google sign in failed ", e);
            }
        }else{
            //Response meant for Facebook.
            //Pass the activity result back to the Facebook SDK
            mCallbackManager.onActivityResult(requestCode, resultCode, data);
        }

    }

    public void saveUserAndRedirect(FirebaseUser loggedInUser){
        try{
            //Save user basic details into shared preferences.
            session.setUserDetails(loggedInUser.getEmail(), loggedInUser.getDisplayName(), String.valueOf(loggedInUser.getPhotoUrl()));
        }catch (NullPointerException e){
            e.printStackTrace();
        }

        startActivity(new Intent(WelcomeScreenActivity.this, ProfilePageActivity.class));
    }
}

File: app/src/main/java/com/iyaselerehoboth/otrek/WelcomeScreenActivity.kt
package com.iyaselerehoboth.otrek

import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.Handler
import android.util.Log
import android.view.View
import android.view.WindowManager
import android.widget.Toast
import androidx.databinding.DataBindingUtil
import com.facebook.AccessToken
import com.facebook.CallbackManager
import com.facebook.FacebookCallback
import com.facebook.FacebookException
import com.facebook.login.LoginManager
import com.facebook.login.LoginResult
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import com.google.android.gms.auth.api.signin.GoogleSignInClient
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.api.ApiException
import com.google.firebase.auth.FacebookAuthProvider
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.GoogleAuthProvider
import com.iyaselerehoboth.otrek.ui.RegisterActivity
import com.iyaselerehoboth.otrek.databinding.ActivityWelcomeScreenBinding
import com.iyaselerehoboth.otrek.ui.HomeActivity

class WelcomeScreenActivity : AppCompatActivity() {

    private lateinit var binding: ActivityWelcomeScreenBinding
    private lateinit var mCallbackManager: CallbackManager
    private lateinit var mAuth: FirebaseAuth
    private lateinit var googleSignInClient: GoogleSignInClient

    companion object{
        private const val RC_SIGN_IN = 9001
        private const val TAG = "OTrek Check"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        //Data binding declaration
        binding = DataBindingUtil.setContentView(this, R.layout.activity_welcome_screen)

        window.setFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS, WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS)

        mAuth = FirebaseAuth.getInstance()

        val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(R.string.default_web_client_id.toString())
                .requestEmail()
                .requestProfile()
                .build()

        googleSignInClient = GoogleSignIn.getClient(this, gso)

        binding.imgbtnFacebook.setOnClickListener { facebookLogin() }
        binding.imgbtnGoogle.setOnClickListener { googleSignIn() }
        binding.btnAlreadyHaveAccount.setOnClickListener { startActivity(Intent(this, RegisterActivity::class.java)) }
        binding.btnSignIn.setOnClickListener { startActivity(Intent(this, HomeActivity::class.java)) }

        //Display the login buttons after 1sec. but need to animate it
        Handler().postDelayed({
            runOnUiThread(
                Runnable {
                    binding.bodyLayout.visibility = View.VISIBLE
                })
        }, 1000)
    }

    override fun onStart() {
        super.onStart()
        val currentUser = mAuth.currentUser
    }

    fun signUserOut(){
        mAuth.signOut()

        //Facebook log out
        LoginManager.getInstance()?.logOut()

        //Google sign out
        googleSignInClient.signOut()?.addOnCompleteListener {
            //Maybe do something
        }
    }

    private fun facebookLogin(){
        //Initialize Facebook Login Button
        mCallbackManager = CallbackManager.Factory.create()
        LoginManager.getInstance().logInWithReadPermissions(this, listOf("email", "public_profile"))
        LoginManager.getInstance().registerCallback(mCallbackManager, object : FacebookCallback<LoginResult>{
            override fun onSuccess(result: LoginResult?) {
                firebaseAuthWithFacebook(result!!.accessToken)
            }

            override fun onCancel() {
                TODO("Not yet implemented")
            }

            override fun onError(error: FacebookException?) {
                Log.d("OTrek Check", "facebook:onError", error)
            }
        })
    }

    private fun googleSignIn(){
        val signInIntent = googleSignInClient.signInIntent
        startActivityForResult(signInIntent, RC_SIGN_IN)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if(requestCode == RC_SIGN_IN){

            val task = GoogleSignIn.getSignedInAccountFromIntent(data)
            try{
                //Sign in was successful, authenticate with firebase
                val account = task?.getResult(ApiException::class.java)
                firebaseAuthWithGoogle(account!!)
            }catch (e: ApiException){
                Log.d(TAG, "Google sign in failed", e)
            }

        }else{

            //Pass the activity result back to the facebook sdk
            mCallbackManager.onActivityResult(requestCode, resultCode, data)
        }

    }

    private fun firebaseAuthWithFacebook(token : AccessToken){
        //Show progress bar here.

        val credential = FacebookAuthProvider.getCredential(token.token)
        mAuth.signInWithCredential(credential)
                .addOnCompleteListener { task ->
                    if(task.isSuccessful){
                        val user = mAuth.currentUser
                    }else{
                        Log.d("OTrek Check", "signInWIthCredential:failure", task.exception)
                        Toast.makeText(this, "Authentication failed", Toast.LENGTH_SHORT).show()
                    }

                    //Hide the progress bar here.
                }
    }

    private fun firebaseAuthWithGoogle(acct: GoogleSignInAccount){
        //Show progress bar here
        val credential = GoogleAuthProvider.getCredential(acct.idToken, null)
        mAuth.signInWithCredential(credential)
                .addOnCompleteListener { task ->
                    if(task.isSuccessful){
                        val user = mAuth.currentUser
                    }else{
                        Log.d(TAG, "signInWithCredential:failure", task.exception)
                    }

                    //Hide progress bar here
                }
    }

}
